packages/ethereum/src/events/index.ts

```diff
@@ -20,85 +20,234 @@ export default class Event {
+                return { ...s, status: Number(status[index].toString()), isSender: true };
             }
+            return { ...s, status: Number(status[index].toString()) };
+        });
+    }
 
-            return result;
-        }, []);
+    async _getPast(_filter?: any, fromBlock?: string | number, toBlock?: string | number) {
+        const swaps: SwapEvent[] = [];
+        const withdraws: WithdrawEvent[] = [];
+        const refunds: RefundEvent[] = [];
 
-        return result;
-    }
+        const logsFilter = {
+            fromBlock: fromBlock || this.config.originBlock, // TODO: constants
+            toBlock: toBlock || 'latest',
+            address: this.config.contractAddress,
+        };
 
-    async subscribe(onMessage: Function, filter?: Function) {
-        this.contract.contract.on('NewContract', (...args: []) => {
-            const swap = TransformNewContract(args);
-            const result = filter ? filter(swap) : swap;
-            if (result) {
-                onMessage(result);
+        const logs = await this.contract.provider.getLogs(logsFilter);
+
+        logs.forEach((log: Log) => {
+            const parsed = this.interface.parseLog(log);
+
+            switch (parsed.name) {
+                case 'NewContract': {
+                    const swap = {
+                        network: 'ETH',
+                        eventName: 'NEW_CONTRACT',
+                        id: parsed.values.id,
+                        hashLock: parsed.values.hashLock,
+                        sender: parsed.values.sender,
+                        receiver: parsed.values.receiver,
+                        inputAmount: parsed.values.inputAmount.toString(),
+                        outputAmount: parsed.values.outputAmount.toString(),
+                        expiration: parsed.values.expiration.toString(),
+                        outputNetwork: parsed.values.outputNetwork,
+                        outputAddress: parsed.values.outputAddress,
+                        transactionHash: log.transactionHash,
+                    };
+
+                    if (_filter.new) {
+                        if (filter(_filter.new, swap)) {
+                            swaps.push(swap);
+                        }
+                    } else {
+                        swaps.push(swap);
+                    }
+
+                    break;
+                }
+
+                case 'Withdraw': {
+                    const withdraw = {
+                        network: 'ETH',
+                        eventName: 'WITHDRAW',
+                        id: parsed.values.id,
+                        secret: parsed.values.secret,
+                        hashLock: parsed.values.hashLock,
+                        sender: parsed.values.sender,
+                        receiver: parsed.values.receiver,
+                        transactionHash: log.transactionHash,
+                    };
+
+                    if (_filter.withdraw) {
```

Comment:
Filter can be extracted in separate function as well. This is how it will look's like:
```
const filterData = (obj, arrayToFill) => {
    obj.withdraw && filter(obj.withdraw, filter) ? arrayToFill.push(withdraw) : arrayToFill.push(withdraw)
}
```
Hmm this if condition,  if (filter withdraw), we push it to the withdraws but if _filter.withdraw is false, again we are pushing to the withdraws array.
Reusable on lines: 104, 127, 149.